### ISSUE TITLE
Sum vs Mean for reconstruction loss

### ISSUE THREAD

**Issue by ctr26** [clementchadebec/benchmark_VAE#124]
This is more a design question. I think what I see in the code is that the reconstruction loss accumlates from the sum MSE. For images of 256 (for me) this means a very large loss term which dwarfs any additional loss component. Specifically in my case in VQVAE the vector quantised loss is ~ 0.1 while the recon loss is 1e4. The thing I've found is that the training process never really brings those numbers close enough to make a difference. As such I've been using the mean reconstruction loss which so far seems to be performing better for me.

Q1. Was this a design decision and is there a strong reason?
Q2. Am I OK to PR so users so they can choose between average and sum?

Thank

CR

**Comment by clementchadebec** [comment]
Hi @ctr26 , 

Thank you for opening this issue. Actually, the rationale for using the L2 loss is that we typically assume that, for images, the decoding distribution (p(x|z)) is a Gaussian distribution with unitary covariance. Following the expression of the ELBO, we want to maximize the log-likelihood of this distribution (*i.e.* log p(x|z), the probability of the input image given a latent code). If you do the math and assume that your decoder only outputs the mean of p(x|z) (constant covariance) you will see that this amounts to minimizing the L2 between the reconstructed images and the input (see for instance [this link](https://stats.stackexchange.com/questions/323568/help-understanding-reconstruction-loss-in-variational-autoencoder)). Nonetheless, you are absolutely right when you say that this is also a design choice. You can indeed, in theory, choose any distribution you want for your decoder to model the data (the main limitation remains that you still end up with a tractable ELBO). Hence, you can, for instance, choose a covariance that is not unitary. In particular, if you set the covariance to be diagonal with coefficients equal to the inverse of the number of pixels, you will end up with the averaged loss that you mentioned. 

So maybe, we can add the possibility to the user to set the value for the covariance? 1 by default and it can be changed in the `ModelConfig`. Let me know your thoughts :)

Best,

Clement

**Comment by ctr26** [comment]
Thanks for the quick response! I have a few more questions just clarify.

a) This might be naive or I may have missed something but doesn't your loss function for VAE (and vq) need an additional sqrt for it to be L2? My understanding of the reduction="none" term means that it returns a list of squares (x-x')**2

    def loss_function(self, recon_x, x):
        MSE = F.mse_loss(
            recon_x.reshape(x.shape[0], -1), x.reshape(x.shape[0], -1), reduction="none"
        ).sum(dim=-1)
        return MSE.mean(dim=0)

In my mind should be:

    def loss_function(self, recon_x, x):
        squares = F.mse_loss(
            recon_x.reshape(x.shape[0], -1), x.reshape(x.shape[0], -1), reduction="none"
        )
        sum_of_squares = squares.sum(dim=-1)
        mse =  squares.mean(dim=-1)
        l2 = torch.sqrt(sum_of_squares)
        rmse = torch.sqrt(mse)
        return l2.mean(dim=0)
        
b) My understanding of VQVAE specifically is that the authors assume a uniform distribute in their latent space. With regard to having the vq model config carry a covariance metric, I'm not completely certain how this would work, nor how accurate it would in this regard? I also am not sure how this affects the loss function practically

b.ii) As a related note: https://github.com/google-deepmind/sonnet/blob/v2/examples/vqvae_example.ipynb, deepmind not only seem to use the true MSE for their reconstruction loss, they also scale that by the image variance (that seems like a data parameter thing though). 

Would love to hear your thoughts.

CR

**Comment by clementchadebec** [comment]
Happy to help :)

a) Actually, the statement that the log-likelihood of the Gaussian distribution amounts to L2 loss was inaccurate. It actually amounts to **squared** L2 loss (or MSE). Sorry about that, I have updated my previous comment. So to me the VAE (and VQ-VAE) losses make sense.

b) As to the covariance I mentioned earlier it is the one in the **decoding** distribution (the one modelled by your decoder) that is in most VAE implementations chosen as a Gaussian with unitary covariance $p_{\theta}(x|z) = N(\mu_{\theta}(z), I)$ where $\mu_{\theta}$ is given by your decoder. I think that what you mention is the **prior** distribution that is actually chosen as a uniform distribution for the VQ-VAE and standard Gaussian for the vanilla VAE. Hence, the covariance to be passed in the `ModelConfig` should reflect the choice for the **decoding** distribution and not the prior.  For instance, if you assume the covariance to be diagonal and whose inverse has 0.1 coefficients on its diagonal the reconstruction loss would become:
```python
recon_loss = 0.1 * 0.5 * F.mse_loss(
   recon_x.reshape(x.shape[0], -1),
   x.reshape(x.shape[0], -1),
   reduction="none",
).sum(dim=-1)
```
Since the Gaussian log-likelihood would become
$\log p_{\theta}(x|z) = -\frac{n}{2} \log (2\pi) - \frac{1}{2} \log|\Sigma| -\frac{1}{2}(x-\mu_{\theta}(z))^{\top}\Sigma^{-1}(x-\mu_{\theta}(z))$
where $x$ is your input data. Note that the 2 first terms are constant with respect to $\mu$. In practice, choosing $\Sigma$ amounts to choosing a scaling parameter for your reconstruction loss what you were ultimately proposing to do.

b.ii) Thank you for bringing this example to my attention. Actually, it seems to me that they are doing exactly what we mentioned earlier meaning choosing a covariance that is different from identity for their decoding distribution. They only decide to use something related to the empirical variance of the input image.

i hope this clarifies things :)

Best,

Clément

**Comment by ctr26** [comment]
Thank you for the very thorough response that has cleared things up.

b) Apologies for sewing confusion there you're absolutely thats it's the prior that is uniform. 

Would it be worth while actually adding reconstruction loss as a function in the base variational auto encoder class and have all the varational models inherit from VAEBase class or are there edge cases ? Currently as I see it theres a good amount of repeated reconstruction_loss code that could be abstracted (happy to do this).

a) on my voyage to clarity I also found this paper

https://arxiv.org/pdf/2006.10273.pdf

Which claims that we should be using log(MSE) for reconstruction loss and that variational loss added to MSE causes over regularisation. 

- What are your thoughts? 
- I'm guessing currently its the negative likelihood being minimized and log(MSE) would be logliklihood?
-  if the author is right then optimising on $L_\text{recon} e^{L_{\text{variational}}}$ would be equivalent, where $L_\text{recon}$ is MSE?

**Comment by clementchadebec** [comment]
Hi @ctr26, and sorry for the late reply,

Thank you for proposing to refactorise some parts of the code. There are indeed some repetitive functions in the VAE classes, especially in the losses. I did that to favor the readability of the code, and since some variants do not use exactly the MSE reconstruction loss (see e.g. IWAE whose loss must also be averaged on the number of importance samples). Nonetheless, you are right, I think we can definitely put the reconstruction loss in the base VAE class and overwrite it in the classes of edge cases anyway :) Do not hesitate to send a PR if you have some time and we can maybe discuss some design choices there as well :)

As to the paper that you shared, thank you very much for the reference. However, I am not sure that the justification for using the log-MSE is quite right. In fact, I found the statement "Equation 42 is minimized when $\sigma^2_{\theta} = MSE$" used to justify the log-MSE loss a bit awkward and I am not sure you can have such conclusions since $\sigma^2_{\theta}$ is assumed to be fixed in the author's paper while the MSE actually depends on $x$. Moreover, I have not seen other implementations or papers using this kind of approach.

Best,

Clément

**Comment by ctr26** [comment]
Yeah, I've given it a

**Comment by clementchadebec** [comment]
Hi @ctr26, happy to see that things are clearer now :) 

I think you are right regarding the commitment loss of the VQVAE model. After having a second look at the paper, you are right this should indeed be a sum and not a mean. Nice catch!

**Comment by ctr26** [comment]
Am I ok to PR this fix?ThanksCROn 16 Jan 2024, at 17:17, Clément Chadebec ***@***.***> wrote:﻿
Hi @ctr26, happy to see that things are clearer now :)
I think you are right regarding the commitment loss of the VQVAE model. After having a second look at the paper, you are right this should indeed be a sum and not a mean. Nice catch!

**Comment by clementchadebec** [comment]
Sure! You can definitely PR this :) 

Thanks!
